### PR TITLE
Notification Queue - multiple GCN notifications

### PR DESCRIPTION
**Pull request by mcoughlin**
This is a follow-on PR to https://github.com/skyportal/skyportal/pull/4043, incorporating the code of https://github.com/skyportal/skyportal/pull/3967 in the new notification framework.

### CODE REVIEW

File: services/notification_queue/notification_queue.py [hunk 1]
```diff
@@ -0,0 +1,1115 @@
...
+
+
+def send_slack_notification(session, target):
+    resource_type = notification_resource_type(target)
+    notifications_prefs = user_preferences(target, "slack", resource_type)
+    if not notifications_prefs:
+        return
+    integration_url = target.user.preferences['slack_integration'].get('url')
+
+    slack_microservice_url = f'http://127.0.0.1:{cfg["slack.microservice_port"]}'
```
Comment (Theodlz) [inline services/notification_queue/notification_queue.py:164]: here we might want to use the app URL and port instead. Not sure how this behaves in a docker-compose network (idk if everything is on localhost or not).

**Comment by Theodlz** [comment]
Very nice overall. It's already "ready", but some minor improvements can be added. I will play around with the code today ingesting lots of things at once (especially skymaps) to see how this behaves. But there is no better test than production to understand how this might behave when a lot of users have lets say email, slack ... and other non SP notifications that trigger external services, introducing delays.

File: services/notification_queue/notification_queue.py (hunk 1 above)
Comment (mcoughlin) [reply services/notification_queue/notification_queue.py:164]: This is also how we do the other microservices.

File: services/notification_queue/notification_queue.py [hunk 2]
```diff
@@ -0,0 +1,1115 @@
...
+                send_email(
+                    recipients=[target.user.contact_email],
+                    subject=subject,
+                    body=body,
+                )
+                print(
+                    f"Sent email notification to user {target.user.id} at email: {target.user.contact_email}, subject: {subject}, body: {body}, resource_type: {resource_type}"
+                )
+            except Exception as e:
+                print(f"Error sending email notification: {e}")
```
[the remark this answers is not in the document]
Comment (mcoughlin) [reply services/notification_queue/notification_queue.py:271]: The log for the microservices is stdout, so this is what is saved.

File: services/notification_queue/notification_queue.py [hunk 3]
```diff
@@ -0,0 +1,1115 @@
...
+
+    if sending:
+        try:
+            message = f"Greetings. This is the SkyPortal robot. {target.text}"
+            client.calls.create(
+                twiml=VoiceResponse().append(Say(message=message)),
+                from_=from_number,
+                to=target.user.contact_phone.e164,
+            )
+            print(
```
[the remark this answers is not in the document]
Comment (mcoughlin) [reply services/notification_queue/notification_queue.py:356]: log == stdout

File: services/notification_queue/notification_queue.py [hunk 4]
```diff
@@ -0,0 +1,1115 @@
...
+    )
+    ws_flow = Flow()
+    ws_flow.push(user_id, "skyportal/FETCH_NOTIFICATIONS")
+
+
+class NotificationQueue(asyncio.Queue):
+    async def service(self):
+        while True:
+
+            notification_id = await queue.get()
```
[the remark this answers is not in the document]
Comment (mcoughlin) [reply services/notification_queue/notification_queue.py:433]: good idea.

File: services/notification_queue/notification_queue.py [hunk 5]
```diff
@@ -0,0 +1,1115 @@
...
+                send_whatsapp_notification(notification)
+                send_email_notification(session, notification)
+                send_slack_notification(session, notification)
+
+
+queue = NotificationQueue()
+
+
+class QueueHandler(tornado.web.RequestHandler):
+    def get(self):
```
[the remark this answers is not in the document]
Comment (mcoughlin) [reply services/notification_queue/notification_queue.py:456]: +1

File: services/notification_queue/notification_queue.py [hunk 6]
```diff
@@ -0,0 +1,1115 @@
...
+                                        text=f"New comment on favorite source *{target_data['obj_id']}*",
+                                        notification_type="favorite_sources_new_comment",
+                                        url=f"/source/{target_data['obj_id']}",
+                                    )
+                                    session.add(notification)
+                                    session.commit()
+                                    await queue.put(notification.id)
+
+            self.set_status(200)
+            return self.write(
```
[the remark this answers is not in the document]
Comment (mcoughlin) [reply services/notification_queue/notification_queue.py:1089]: I am going to set a 2 second timeout for now. I think rethinking the queue could be a thing in a future PR

File: skyportal/models/user_notification.py [hunk 7]
```diff
@@ -502,532 +109,18 @@ def receive_after_commit(session):
 
         IOLoop.current().run_in_executor(
```
[the remark this answers is not in the document]
Comment (mcoughlin) [reply skyportal/models/user_notification.py:110]: good

File: services/notification_queue/notification_queue.py [hunk 8]
```diff
@@ -0,0 +1,1115 @@
...
+
+
+def send_whatsapp_notification(target):
+    resource_type = notification_resource_type(target)
+    prefs = user_preferences(target, "whatsapp", resource_type)
+    if not prefs:
+        return
+
+    sending = False
+    if prefs[resource_type]['whatsapp'].get("on_shift", False):
```
[the remark this answers is not in the document]
Comment (mcoughlin) [reply services/notification_queue/notification_queue.py:370]: future PR

File: services/notification_queue/notification_queue.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Theodlz) [reply services/notification_queue/notification_queue.py:356]: Doesnt log provide the nice display with the make log with timestamp and so on? Or does that also come with supervisor stdout?

[the remark this answers is not in the document]
Comment (mcoughlin) [reply services/notification_queue/notification_queue.py]: @Theodlz good point. I pushed a log.

**Pull request merged**